compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributedInterfaceSupertypeGenerator.kt

```diff
@@ -215,6 +220,87 @@ internal class ContributedInterfaceSupertypeGenerator(
       // TODO warn?
     }
 
+    if (session.metroFirBuiltIns.options.enableDaggerAnvilInterop) {
+      val unmatchedRankReplacements = mutableSetOf<ClassId>()
+      val pendingRankReplacements = processRankBasedReplacements(contributions, typeResolver)
+
+      pendingRankReplacements.distinct().forEach { replacedClassId ->
+        val removed = contributions.remove(replacedClassId)
+        if (removed != null) {
+          unmatchedRankReplacements += replacedClassId
+        }
+      }
+
+      if (unmatchedRankReplacements.isNotEmpty()) {
+        // TODO we could report all rank based replacements here
+      }
+    }
+
     return contributions.values.toList()
   }
+
+  /**
+   * This is an imperfect solution to provide `rank` interop for users migrating from Dagger-Anvil.
+   * We're not able to get 1:1 parity due to some type restrictions but it should be enough to make
+   * the migration much more feasible in large projects that have a lot of ranked bindings.
+   *
+   * There are two important limitations to note here:
+   * 1. Ranked bindings and bindings that get outranked must both explicitly declare their binding
+   *    type in order for us to actually compare them, because supertypes are not resolvable here.
+   *    The user will end up getting a duplicate binding error if the outranked binding is using an
+   *    implicit type.
+   * 2. We can't check for qualifiers when comparing these bindings because those annotation calls
+   *    are not resolved at this point. E.g. a compiler critical annotation
+   *    like @ContributesBinding(..) will be resolved but @Named(..) will not. This means that
+   *    qualifiers are unsupported for rank interop support. Rank can only effectively be used for
+   *    binding types where all of them are unqualified or all use the same qualifier. Other
+   *    combinations will need to be migrated to instead use explicit replacements or exclusions.
```

Comment:
This may not be a true limitation. You can create use DeclarationPredicate to define meta-annotations.
https://github.com/JetBrains/kotlin/blob/52a3ec9184fa44b2c8ce981f279cd66686dbe73b/plugins/plugin-sandbox/src/org/jetbrains/kotlin/plugin/sandbox/fir/SomeAdditionalSupertypeGenerator.kt#L28-L33

This should cause anything annotated with the meta-annotation to also become a "compiler critical annotation". (Registering the predicates is what causes the annotations to become "critical", meta-predicates just automatically register anything they annotate)

Reply:
Iiiinteresting. Do you think that would support this case to? https://github.com/ZacSweers/metro/issues/205

Reply:
I may not fully understand what the problems with #205 where, but I don't think so. I don't think it will cause the class references to `ContributesTo`-annotated interfaces to become resolved. I suspect it only impacts the resolution of annotations that are annotated.

But you could certainly try it! I don't have direct knowledge about what to expect.

Reply:
You could hard-pivot and require all scopes to be annotation classes! 😅

Reply:
Wow I didn't know there were mechanisms for meta annotations. Will give this a try once I have a chance today!

Reply:
[Meta-meta-meta-annotations too!](https://github.com/JetBrains/kotlin/blob/adc0a0eb422ad7cca2be8fe798b2d5383a3a27ce/plugins/plugin-sandbox/testData/diagnostics/supertypes/metaAnnotationOrder.kt#L34-L45)